File: Liftr/BodyWeightClient.swift
import Foundation
import Supabase

enum BodyWeightSource: String, Codable, CaseIterable, Hashable {
    case manual
    case appleHealth = "apple_health"
    case healthConnect = "health_connect"
}

struct BodyWeightEntry: Decodable, Identifiable, Hashable {
    let id: UUID
    let user_id: UUID
    let measured_at: Date
    let weight_kg: Double
    let source: String
    let external_sample_id: String?
    let created_at: Date?
    let updated_at: Date?

    var sourceKind: BodyWeightSource {
        BodyWeightSource(rawValue: source) ?? .manual
    }
}

struct BodyWeightUpsertResult: Decodable, Hashable {
    let entry_id: UUID?
    let inserted: Bool?
    let duplicate: Bool?
}

struct BodyWeightChartPoint: Identifiable, Hashable {
    let id: UUID
    let label: String
    let value: Double
    let measuredAt: Date
}

enum BodyWeightRangePreset: String, CaseIterable, Identifiable {
    case days30
    case days90
    case days365

    var id: String { rawValue }

    var title: String {
        switch self {
        case .days30: return "30 days"
        case .days90: return "90 days"
        case .days365: return "365 days"
        }
    }

    func startDate(relativeTo now: Date = Date()) -> Date {
        let days: Int
        switch self {
        case .days30: days = 30
        case .days90: days = 90
        case .days365: days = 365
        }
        return Calendar.current.date(byAdding: .day, value: -days, to: now) ?? now
    }
}

enum BodyWeightPresentation {
    static func formatKg(_ value: Double) -> String {
        String(format: "%.1f kg", value)
    }

    static func deltaText(current: Double, previous: Double?) -> String? {
        guard let previous else { return nil }
        let delta = current - previous
        if abs(delta) < 0.05 { return "No change vs previous entry" }
        let sign = delta > 0 ? "+" : ""
        return "\(sign)\(String(format: "%.1f", delta)) kg vs previous entry"
    }

    static func periodDeltaText(entries: [BodyWeightEntry], days: Int, now: Date = Date()) -> String? {
        let sorted = entries.sorted { $0.measured_at < $1.measured_at }
        guard let latest = sorted.last else { return nil }
        guard let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: now) else { return nil }
        let baseline = sorted.last(where: { $0.measured_at <= cutoff }) ?? sorted.first
        guard let baseline, baseline.id != latest.id || baseline.measured_at <= cutoff else { return nil }
        let delta = latest.weight_kg - baseline.weight_kg
        if abs(delta) < 0.05 { return "No net change in the last \(days) days" }
        let sign = delta > 0 ? "+" : ""
        return "\(sign)\(String(format: "%.1f", delta)) kg in the last \(days) days"
    }

    static func chartPoints(from entries: [BodyWeightEntry], preset: BodyWeightRangePreset, now: Date = Date()) -> [BodyWeightChartPoint] {
        let start = preset.startDate(relativeTo: now)
        let filtered = entries
            .filter { $0.measured_at >= start }
            .sorted { $0.measured_at < $1.measured_at }
        let formatter = DateFormatter()
        formatter.dateStyle = .short
        formatter.timeStyle = .none
        return filtered.map {
            BodyWeightChartPoint(
                id: $0.id,
                label: formatter.string(from: $0.measured_at),
                value: $0.weight_kg,
                measuredAt: $0.measured_at
            )
        }
    }

    static func sourceLabel(_ source: BodyWeightSource) -> String {
        switch source {
        case .manual: return "Manual"
        case .appleHealth: return "Apple Health"
        case .healthConnect: return "Health Connect"
        }
    }
}

enum BodyWeightClient {
    private struct UpsertParams: Encodable {
        let p_measured_at: String
        let p_weight_kg: Double
        let p_source: String
        let p_external_sample_id: String?
    }

    private static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    static func listEntries(limit: Int = 500) async throws -> [BodyWeightEntry] {
        let res = try await SupabaseManager.shared.client
            .from("body_weight_entries")
            .select("id,user_id,measured_at,weight_kg,source,external_sample_id,created_at,updated_at")
            .order("measured_at", ascending: false)
            .limit(limit)
            .execute()
        return try JSONDecoder.supabase().decode([BodyWeightEntry].self, from: res.data)
    }

    static func upsertEntry(
        measuredAt: Date,
        weightKg: Double,
        source: BodyWeightSource,
        externalSampleId: String? = nil
    ) async throws -> BodyWeightUpsertResult {
        let params = UpsertParams(
            p_measured_at: iso.string(from: measuredAt),
            p_weight_kg: weightKg,
            p_source: source.rawValue,
            p_external_sample_id: externalSampleId
        )
        let res = try await SupabaseManager.shared.client
            .rpc("upsert_body_weight_entry", params: params)
            .execute()
        return try JSONDecoder.supabase().decode(BodyWeightUpsertResult.self, from: res.data)
    }

    static func deleteEntry(id: UUID) async throws {
        _ = try await SupabaseManager.shared.client
            .from("body_weight_entries")
            .delete()
            .eq("id", value: id.uuidString)
            .execute()
    }

    static func updateManualEntry(id: UUID, measuredAt: Date, weightKg: Double) async throws {
        let payload: [String: AnyEncodable] = [
            "measured_at": AnyEncodable(iso.string(from: measuredAt)),
            "weight_kg": AnyEncodable(weightKg),
            "source": AnyEncodable(BodyWeightSource.manual.rawValue)
        ]
        _ = try await SupabaseManager.shared.client
            .from("body_weight_entries")
            .update(payload)
            .eq("id", value: id.uuidString)
            .execute()
    }
}

File: android/app/src/main/java/com/lilru/liftr/bodyweight/BodyWeightClient.kt
package com.lilru.liftr.bodyweight

import com.lilru.liftr.data.BackendContracts
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlin.math.abs

enum class BodyWeightSource(val wire: String) {
    Manual("manual"),
    AppleHealth("apple_health"),
    HealthConnect("health_connect");

    companion object {
        fun fromWire(value: String): BodyWeightSource =
            entries.firstOrNull { it.wire == value } ?: Manual
    }
}

@Serializable
data class BodyWeightEntryWire(
    val id: String,
    @SerialName("user_id") val userId: String,
    @SerialName("measured_at") val measuredAt: String,
    @SerialName("weight_kg") val weightKg: Double,
    val source: String,
    @SerialName("external_sample_id") val externalSampleId: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null
)

@Serializable
data class BodyWeightUpsertResultWire(
    @SerialName("entry_id") val entryId: String? = null,
    val inserted: Boolean? = null,
    val duplicate: Boolean? = null
)

data class BodyWeightChartPoint(
    val id: String,
    val label: String,
    val value: Double,
    val measuredAt: Instant
)

enum class BodyWeightRangePreset(val title: String, val days: Int) {
    Days30("30 days", 30),
    Days90("90 days", 90),
    Days365("365 days", 365)
}

object BodyWeightPresentation {
    fun formatKg(value: Double): String = String.format(Locale.US, "%.1f kg", value)

    fun deltaText(current: Double, previous: Double?): String? {
        if (previous == null) return null
        val delta = current - previous
        if (abs(delta) < 0.05) return "No change vs previous entry"
        val sign = if (delta > 0) "+" else ""
        return "$sign${String.format(Locale.US, "%.1f", delta)} kg vs previous entry"
    }

    fun periodDeltaText(entries: List<BodyWeightEntryWire>, days: Int, now: Instant = Instant.now()): String? {
        val sorted = entries.sortedBy { it.measuredAt }
        val latest = sorted.lastOrNull() ?: return null
        val cutoff = now.minusSeconds(days.toLong() * 86_400L)
        val baseline = sorted.lastOrNull { Instant.parse(it.measuredAt) <= cutoff } ?: sorted.firstOrNull()
        if (baseline == null) return null
        val delta = latest.weightKg - baseline.weightKg
        if (abs(delta) < 0.05) return "No net change in the last $days days"
        val sign = if (delta > 0) "+" else ""
        return "$sign${String.format(Locale.US, "%.1f", delta)} kg in the last $days days"
    }

    fun chartPoints(entries: List<BodyWeightEntryWire>, preset: BodyWeightRangePreset, now: Instant = Instant.now()): List<BodyWeightChartPoint> {
        val start = now.minusSeconds(preset.days.toLong() * 86_400L)
        val formatter = DateTimeFormatter.ofPattern("d MMM").withLocale(Locale.getDefault())
        return entries
            .mapNotNull { entry ->
                val instant = runCatching { Instant.parse(entry.measuredAt) }.getOrNull() ?: return@mapNotNull null
                if (instant.isBefore(start)) return@mapNotNull null
                BodyWeightChartPoint(
                    id = entry.id,
                    label = formatter.format(instant.atZone(ZoneOffset.systemDefault())),
                    value = entry.weightKg,
                    measuredAt = instant
                )
            }
            .sortedBy { it.measuredAt }
    }

    fun sourceLabel(source: BodyWeightSource): String = when (source) {
        BodyWeightSource.Manual -> "Manual"
        BodyWeightSource.AppleHealth -> "Apple Health"
        BodyWeightSource.HealthConnect -> "Health Connect"
    }
}

class BodyWeightClient(
    private val supabase: SupabaseClient
) {
    suspend fun listEntries(limit: Int = 500): List<BodyWeightEntryWire> {
        return supabase.from(BackendContracts.Tables.BODY_WEIGHT_ENTRIES).select {
            order(column = "measured_at", order = Order.DESCENDING)
            limit(count = limit.toLong())
        }.decodeList<BodyWeightEntryWire>()
    }

    suspend fun upsertEntry(
        measuredAt: Instant,
        weightKg: Double,
        source: BodyWeightSource,
        externalSampleId: String? = null
    ): BodyWeightUpsertResultWire {
        val params = buildJsonObject {
            put("p_measured_at", JsonPrimitive(measuredAt.toString()))
            put("p_weight_kg", JsonPrimitive(weightKg))
            put("p_source", JsonPrimitive(source.wire))
            externalSampleId?.let { put("p_external_sample_id", JsonPrimitive(it)) }
        }
        return supabase.postgrest.rpc(BackendContracts.Rpc.UPSERT_BODY_WEIGHT_ENTRY, params)
            .decodeAs<BodyWeightUpsertResultWire>()
    }

    suspend fun deleteEntry(id: String) {
        supabase.from(BackendContracts.Tables.BODY_WEIGHT_ENTRIES).delete {
            filter { eq("id", id) }
        }
    }
}
